src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -630,12 +650,44 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
                     Debug.Assert(_inputBuffer != null);
                     // In this case, we don't have an entire block's worth left, so store it up in the
                     // input buffer, which by now must be empty.
-                    Buffer.BlockCopy(buffer, currentInputIndex, _inputBuffer, 0, bytesToWrite);
+                    buffer.Slice(currentInputIndex, bytesToWrite).CopyTo(_inputBuffer);
                     _inputBufferIndex += bytesToWrite;
                     return;
                 }
             }
             return;
+
+            unsafe static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
+            {
+                if (MemoryMarshal.TryGetArray(inputBuffer, out ArraySegment<byte> segment))
+                {
+                    // Skip the copy if readonlymemory is actually an array.
+                    Debug.Assert(segment.Array is not null);
+                    return transform.TransformBlock(segment.Array, segment.Offset, inputBuffer.Length, outputBuffer, outputOffset);
+                }
+                else
+                {
+                    var rentedBuffer = ArrayPool<byte>.Shared.Rent(inputBuffer.Length);
+                    int result = default;
+                    try
+                    {
+                        // Pin the rented buffer for security.
+                        fixed (byte* _ = &rentedBuffer[0])
+                        {
+                            inputBuffer.CopyTo(rentedBuffer);
+                            result = transform.TransformBlock(rentedBuffer, 0, inputBuffer.Length, outputBuffer, outputOffset);
+                        }
+                    }
+                    finally
+                    {
+                        CryptographicOperations.ZeroMemory(rentedBuffer.AsSpan(0, inputBuffer.Length));
```

Comment:
If the array is being pinned for the operation that's presumably so that it can be populated and cleared without GC compaction applying, so I'd expect the clear to be inside the pin.